Don't print two different types identically in an error message

Issue #973 shows some Checker Framework output:

```
SuperConsumer.java:10: error: [argument.type.incompatible] incompatible types in argument.
  found   : List<Object>
  required: ? extends Object super List<Object>
```

Note that there are two occurrences of `List<Object>` in the output.

Running the test case with the `-AprintAllQualifiers` command-line option gives a less confusing error messages, because it becomes clear that the two different occurrences of `List<Object>` in the error message are actually two different types:

```
SuperConsumer.java:10: error: [argument.type.incompatible] incompatible types in argument.
        consumer.accept(list);
                        ^
  found   : @UnknownKeyFor List<@UnknownKeyFor Object>
  required: ? extends @UnknownKeyFor Object super @KeyForBottom List<@KeyForBottom Object>
```

The Checker Framework should never print an error message in which two different types have the same printed representation.
 * One way to do this would be to always print types under the assumption of standard defaulting, even though the Checker Framework uses slightly different defaulting rules in different locations.
 * Another way to do this would be to scan all the types in the error message and turn on `-AprintAllQualifiers` or `-AprintVerboseGenerics` if there are any two different types with the same printed representation.
 * There may be other possible solutions as well.

--- a/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -1840,14 +1840,10 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
      * Return whether or not the verbose toString should be used when printing the two annotated
      * types.
      *
-     * <p>If any simple toString of any annotated type in the first atm or second atm is equal to
-     * any toString of any annotated type in the first or second atm and if the verbose toStrings
-     * differ, then the verbose toString should be used.
-     *
-     * @param atm1 the first atm
-     * @param atm2 the second atm
-     * @return whether or not the verbose toString should be used when printing the two annotated
-     *     types.
+     * @param atm1 the first AnnotatedTypeMirror
+     * @param atm2 the second AnnotatedTypeMirror
+     * @return true iff there are two annotated types (in either ATM) such that their toStrings are
+     *     the same but their verbose toStrings differ
      */
     private boolean shouldPrintVerbose(AnnotatedTypeMirror atm1, AnnotatedTypeMirror atm2) {
         String atm1ToString = atm1.toString();
@@ -1861,6 +1857,7 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
 
         SimpleAnnotatedTypeScanner<Boolean, Void> checkForMismatchedToStrings =
                 new SimpleAnnotatedTypeScanner<Boolean, Void>() {
+                    /** Maps from a type's toString to its verbose toString */
                     Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
 
                     @Override
@@ -1906,7 +1903,7 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
         if (result1 != null && result1) {
             return true;
         }
-        // Call reset to clear the visitor history, but not the map for Strings to types.
+        // Call reset to clear the visitor history, but not the map
         checkForMismatchedToStrings.reset();
         Boolean result2 = checkForMismatchedToStrings.visit(atm2);
         return result2 == null ? false : result2;

--- a/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -1862,17 +1862,9 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
 
                     @Override
                     protected Boolean reduce(Boolean r1, Boolean r2) {
-                        if (r1 == null) {
-                            if (r2 == null) {
-                                return false;
-                            } else {
-                                return r2;
-                            }
-                        } else if (r2 == null) {
-                            return r1;
-                        } else {
-                            return r1 || r2;
-                        }
+                        r1 = r1 == null ? false : r1;
+                        r2 = r2 == null ? false : r2;
+                        return r1 || r2;
                     }
 
                     @Override
@@ -1891,15 +1883,16 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
                     }
                 };
         Boolean r1 = checkForMismatchedToStrings.visit(atm1);
+        if (r1 != null && r1) {
+            return true;
+        }
         // Call reset to clear the visitor history, but not the map from Strings to types.
         checkForMismatchedToStrings.reset();
         Boolean r2 = checkForMismatchedToStrings.visit(atm2);
 
         // SimpleAnnotatedTypeScanner#scan returns null if it encounters a null AnnotatedTypeMirror.
         // This shouldn't happen if the atm1 and atm2 are well-formed.
-        if (r1 == null) r1 = false;
-        if (r2 == null) r2 = false;
-        return r1 || r2;
+        return r2 == null ? false : r2;
     }
 
     protected void checkArrayInitialization(

--- a/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -100,6 +100,7 @@ import org.checkerframework.framework.type.GenericAnnotatedTypeFactory;
 import org.checkerframework.framework.type.QualifierHierarchy;
 import org.checkerframework.framework.type.TypeHierarchy;
 import org.checkerframework.framework.type.VisitorState;
+import org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeScanner;
 import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.framework.util.ContractsUtils;
 import org.checkerframework.framework.util.ContractsUtils.ConditionalPostcondition;
@@ -1822,17 +1823,98 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
 
         // Use an error key only if it's overridden by a checker.
         if (!success) {
-            String valueTypeString = valueType.toString();
-            String varTypeString = varType.toString();
-            // If both types as strings are the same, try printing verbosely.
-            if (valueTypeString.equals(varTypeString)
-                    // or if neither string contains an annotation
-                    || (!valueTypeString.contains("@") && !varTypeString.contains("@"))) {
+            String valueTypeString;
+            String varTypeString;
+            if (shouldPrintVerbose(varType, valueType)) {
                 valueTypeString = valueType.toString(true);
                 varTypeString = varType.toString(true);
+            } else {
+                valueTypeString = valueType.toString();
+                varTypeString = varType.toString();
             }
             checker.report(Result.failure(errorKey, valueTypeString, varTypeString), valueTree);
         }
+    }
+
+    /**
+     * Return whether or not the verbose toString should be used when printing the two annotated
+     * types.
+     *
+     * <p>If any simple toString of any annotated type in the first atm is equal to any toString of
+     * any annotated type in the second atm and if the verbose toStrings differ, then the verbose
+     * toString should be used.
+     *
+     * @param atm1 the first atm
+     * @param atm2 the second atm
+     * @return whether or not the verbose toString should be used when printing the two annotated
+     *     types.
+     */
+    private boolean shouldPrintVerbose(AnnotatedTypeMirror atm1, AnnotatedTypeMirror atm2) {
+        String atm1ToString = atm1.toString();
+        String atm2ToString = atm2.toString();
+        // If both types as strings are the same, use verbose toString.
+        if (atm2ToString.equals(atm1ToString)
+                // or if neither string contains an annotation
+                || (!atm2ToString.contains("@") && !atm1ToString.contains("@"))) {
+            return true;
+        }
+
+        // Collect the "toString" of every AnnotatedTypeMirror in atm2.
+        SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void> collectToStrings =
+                new SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void>() {
+                    Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
+
+                    @Override
+                    protected Map<String, List<AnnotatedTypeMirror>> defaultAction(
+                            AnnotatedTypeMirror type, Void aVoid) {
+                        if (type == null) {
+                            return map;
+                        }
+                        String simple = type.toString();
+                        List<AnnotatedTypeMirror> list;
+                        if (map.containsKey(simple)) {
+                            list = map.get(simple);
+                        } else {
+                            list = new ArrayList<>();
+                            map.put(simple, list);
+                        }
+                        list.add(type);
+                        return map;
+                    }
+                };
+        Map<String, List<AnnotatedTypeMirror>> map = collectToStrings.visit(atm2, null);
+
+        // Compare the "toString" of every AnnotatedTypeMirror in atm2 with the "toString" of
+        // every AnnotatedType mirror in atm1.  If the toStrings match, but toString verbose of
+        // the annotated types does not match, then the types should be printed verbosely.
+        SimpleAnnotatedTypeScanner<Boolean, Map<String, List<AnnotatedTypeMirror>>>
+                checkToStringOfOther =
+                        new SimpleAnnotatedTypeScanner<
+                                Boolean, Map<String, List<AnnotatedTypeMirror>>>() {
+                            @Override
+                            protected Boolean defaultAction(
+                                    AnnotatedTypeMirror type,
+                                    Map<String, List<AnnotatedTypeMirror>> map) {
+                                if (type == null) {
+                                    return false;
+                                }
+                                String simple = type.toString();
+                                if (!map.containsKey(simple)) {
+                                    return false;
+                                }
+                                String verbose = type.toString(true);
+                                List<AnnotatedTypeMirror> list = map.get(simple);
+                                for (AnnotatedTypeMirror other : list) {
+                                    String otherVerbose = other.toString(true);
+                                    if (!verbose.equals(otherVerbose)) {
+                                        return true;
+                                    }
+                                }
+                                return false;
+                            }
+                        };
+        Boolean result = checkToStringOfOther.visit(atm1, map);
+        return result == null ? false : result;
     }
 
     protected void checkArrayInitialization(

--- a/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -1840,9 +1840,9 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
      * Return whether or not the verbose toString should be used when printing the two annotated
      * types.
      *
-     * <p>If any simple toString of any annotated type in the first atm is equal to any toString of
-     * any annotated type in the second atm and if the verbose toStrings differ, then the verbose
-     * toString should be used.
+     * <p>If any simple toString of any annotated type in the first atm or second atm is equal to
+     * any toString of any annotated type in the first or second atm and if the verbose toStrings
+     * differ, then the verbose toString should be used.
      *
      * @param atm1 the first atm
      * @param atm2 the second atm
@@ -1859,62 +1859,57 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
             return true;
         }
 
-        // Collect the "toString" of every AnnotatedTypeMirror in atm2.
-        SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void> collectToStrings =
-                new SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void>() {
+        SimpleAnnotatedTypeScanner<Boolean, Void> checkForMismatchedToStrings =
+                new SimpleAnnotatedTypeScanner<Boolean, Void>() {
                     Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
 
                     @Override
-                    protected Map<String, List<AnnotatedTypeMirror>> defaultAction(
-                            AnnotatedTypeMirror type, Void aVoid) {
+                    protected Boolean reduce(Boolean r1, Boolean r2) {
+                        if (r1 == null) {
+                            if (r2 == null) {
+                                return false;
+                            } else {
+                                return r2;
+                            }
+                        } else if (r2 == null) {
+                            return r1;
+                        } else {
+                            return r1 || r2;
+                        }
+                    }
+
+                    @Override
+                    protected Boolean defaultAction(AnnotatedTypeMirror type, Void avoid) {
                         if (type == null) {
-                            return map;
+                            return false;
                         }
                         String simple = type.toString();
-                        List<AnnotatedTypeMirror> list;
-                        if (map.containsKey(simple)) {
-                            list = map.get(simple);
-                        } else {
+                        List<AnnotatedTypeMirror> list = map.get(simple);
+                        if (list == null) {
                             list = new ArrayList<>();
                             map.put(simple, list);
                         }
+                        String verbose = type.toString(true);
+                        for (AnnotatedTypeMirror other : list) {
+                            String otherVerbose = other.toString(true);
+                            if (!verbose.equals(otherVerbose)) {
+                                list.add(type);
+                                return true;
+                            }
+                        }
                         list.add(type);
-                        return map;
+                        return false;
                     }
                 };
-        Map<String, List<AnnotatedTypeMirror>> map = collectToStrings.visit(atm2, null);
-
-        // Compare the "toString" of every AnnotatedTypeMirror in atm2 with the "toString" of
-        // every AnnotatedType mirror in atm1.  If the toStrings match, but toString verbose of
-        // the annotated types does not match, then the types should be printed verbosely.
-        SimpleAnnotatedTypeScanner<Boolean, Map<String, List<AnnotatedTypeMirror>>>
-                checkToStringOfOther =
-                        new SimpleAnnotatedTypeScanner<
-                                Boolean, Map<String, List<AnnotatedTypeMirror>>>() {
-                            @Override
-                            protected Boolean defaultAction(
-                                    AnnotatedTypeMirror type,
-                                    Map<String, List<AnnotatedTypeMirror>> map) {
-                                if (type == null) {
-                                    return false;
-                                }
-                                String simple = type.toString();
-                                if (!map.containsKey(simple)) {
-                                    return false;
-                                }
-                                String verbose = type.toString(true);
-                                List<AnnotatedTypeMirror> list = map.get(simple);
-                                for (AnnotatedTypeMirror other : list) {
-                                    String otherVerbose = other.toString(true);
-                                    if (!verbose.equals(otherVerbose)) {
-                                        return true;
-                                    }
-                                }
-                                return false;
-                            }
-                        };
-        Boolean result = checkToStringOfOther.visit(atm1, map);
-        return result == null ? false : result;
+        Boolean result1 = checkForMismatchedToStrings.visit(atm1);
+        // SimpleAnnotatedTypeScanner#scan returns null is it encounters a null AnnotatedTypeMirror
+        if (result1 != null && result1) {
+            return true;
+        }
+        // Call reset to clear the visitor history, but not the map for Strings to types.
+        checkForMismatchedToStrings.reset();
+        Boolean result2 = checkForMismatchedToStrings.visit(atm2);
+        return result2 == null ? false : result2;
     }
 
     protected void checkArrayInitialization(

--- a/framework/src/org/checkerframework/framework/type/visitor/SimpleAnnotatedTypeScanner.java
+++ b/framework/src/org/checkerframework/framework/type/visitor/SimpleAnnotatedTypeScanner.java
@@ -26,8 +26,8 @@ public class SimpleAnnotatedTypeScanner<R, P> extends AnnotatedTypeScanner<R, P>
      */
     @Override
     public final R visitDeclared(AnnotatedDeclaredType type, P p) {
-        defaultAction(type, p);
-        return super.visitDeclared(type, p);
+        R r = defaultAction(type, p);
+        return reduce(super.visitDeclared(type, p), r);
     }
 
     /**
@@ -39,8 +39,8 @@ public class SimpleAnnotatedTypeScanner<R, P> extends AnnotatedTypeScanner<R, P>
      */
     @Override
     public final R visitExecutable(AnnotatedExecutableType type, P p) {
-        defaultAction(type, p);
-        return super.visitExecutable(type, p);
+        R r = defaultAction(type, p);
+        return reduce(super.visitExecutable(type, p), r);
     }
 
     /**
@@ -52,8 +52,8 @@ public class SimpleAnnotatedTypeScanner<R, P> extends AnnotatedTypeScanner<R, P>
      */
     @Override
     public final R visitArray(AnnotatedArrayType type, P p) {
-        defaultAction(type, p);
-        return super.visitArray(type, p);
+        R r = defaultAction(type, p);
+        return reduce(super.visitArray(type, p), r);
     }
 
     /**
@@ -65,8 +65,8 @@ public class SimpleAnnotatedTypeScanner<R, P> extends AnnotatedTypeScanner<R, P>
      */
     @Override
     public final R visitTypeVariable(AnnotatedTypeVariable type, P p) {
-        defaultAction(type, p);
-        return super.visitTypeVariable(type, p);
+        R r = defaultAction(type, p);
+        return reduce(super.visitTypeVariable(type, p), r);
     }
 
     /**
@@ -114,7 +114,7 @@ public class SimpleAnnotatedTypeScanner<R, P> extends AnnotatedTypeScanner<R, P>
      */
     @Override
     public final R visitWildcard(AnnotatedWildcardType type, P p) {
-        defaultAction(type, p);
-        return super.visitWildcard(type, p);
+        R r = defaultAction(type, p);
+        return reduce(super.visitWildcard(type, p), r);
     }
 }

